### DISCUSSION TITLE
Todo.txt filter format

### DISCUSSION THREAD

**Discussion by harshad1** [gsantner/markor#1460] [Ideas]
To recap the existing discussion from #859 

@harshad1 :
> @gsantner As a next stage for todo filtering, I would like to implement a simple query language.
> 
> For example (*A | *B) & #work would mean select tasks with priority (A or B) and which have project #work. This would be > loosely inspired by the todoist filter language.
> 
> I want to do this in order to
> 
> Enable more complex filters
> Allow filters to be saved as shortcuts in the 'Browse todo' dialog
> any opinions / thoughts?

@gsantner :
> I can only suggest to not reinvent what is pretty known out there, and java has good support for --- regex search
> ^[A|B].*#work.*

@harshad :
> My current thinking is a json structured language
> 
> {
>     "note": "Todo text filter in json format",
>     "and": [
>         "project": "work",
>         "or" : [
>             "context": "meeting",
>             "context": "george"
>         ],
>         "not" : {
>             "or": [
>                 "priority": "C",
>                 "priority": "D",
>                 "priority": "E"
>             ]
>         },
>         "or": [
>             "due": "today",
>             "due": "overdue"
>         ],
>         "done": false
>     ]
> }
> This will enable better filtering than a regex (must have correct project order etc in regex).
> 
> Each block is roughly like an s-expression.
> 
> Since it is json, we will not need to write our own parser. Just a simple system to translate the json into a query (which is already handled by the todo txt filter dialog)

@gsantner :
> but, who would spontaneously start typing such things..just to get some other representation of the text? It all looks more complex than it should be, or todo.txt itself is.
> 
> The goal of this is to have a implementation of todo.txt for view mode. currently it's just the text shown "as is" in in html.

@harshad1 :
> The idea would be to have one or more saved filters. Then one click -> filtered tasks. What I have above is an example of what I would check in the morning before work.
> 
> It's harder to type than something like
> 
> (+work & ( @meeting | @george ) & !( *C | *D | *E ) & ( today | overdue ) & !done)
> But with the latter we will have to write our own parser.
> 
> > The goal of this is to have a implementation of todo.txt for view mode. currently it's just the text shown "as is" in in html.
> 
> Yeah, this may not be the best place to have this discussion. I can move it to a discussion

@gsantner :
> Yes ofc discussion is fine. But we don't need such a complex thing. Just all tasks with +work is fine. Or just all tasks with priority A.

**Comment by harshad1** [comment]
@gsantner

> But we don't need such a complex thing. Just all tasks with +work is fine. Or just all tasks with priority A.

It would be very useful, for example, to filter 'high priority work tasks which are overdue'

**Reply by gsantner** [reply]
But how much/often do you or Markor users expect this to happen? Just thinking loud: If advanced task management is needed, todo.txt and Markor might just be not the right tool for it.

And if it's really this one specific required browsing option which would be important, we can always add it directly - without any custom query language that would be specific to this one thing.

**Comment by harshad1** [comment]
Looking at my usage over the past day, the 80/20 solution would be to allow filters to be saved. 

So from a filtered view a user could press save and the current view would appear as a shortcut in the filter menu. The user would be prompted to provide a name. Max saved filter count would be 5 from most recent to oldest. 

In the future we could then look into adding an option to chain filters.

This is somewhat less flexible, but would be 100% UI driven and discoverable. Any serialization format would be completely internal.
